### PR TITLE
[train][release] Add async checkpointing/validation benchmark

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

This PR creates a benchmark with the following setup:
* TorchTrainer with VisionTransformer on CIFAR-10 dataset
* Training on 2 gpus and validation on 2 gpus. Cluster can scale accordingly from 2-4 gpus. I considered making this cpu only but figured that moving tensors between cpu and gpu is a big part of the cost of async checkpointing/validation.

The benchmark compares the following checkpointing/validation combos:
1) Synchronous checkpointing + validation within the training TorchTrainer
2) Asynchronous checkpointing + validation with a validation-only TorchTrainer
3) Asynchronous checkpointing + validation with map_batches

The following should be true when comparing these combos:
* e2e time: 2/3 should be slightly faster than 1
* final validation waiting time: 1 should be almost 0 and 2/3 should be fast (not much more than average validation time)
* total report blocked time: 2/3 should be negligible (sometimes just waiting for checkpoint upload to finish) and 1 should be somewhat high (checkpoint upload time)
* total validation time: 2/3 should be slightly higher than 1 because they need to set up new actors
* final score: these should all be similar since they are calculating the same thing

# Testing

[Release test](https://console.anyscale-staging.com/cld_kvedZWag2qA8i5BjxUevf5i7/prj_92c7b71w55flm6gv6imv4m6vqg/jobs/prodjob_yqbe6c8e4gulu5dbp4hybvnfgs?job-logs-section-tabs=application_logs&job-tab=overview) produced good results 

| Train Run Type | e2e time | final validation waiting time (s) | total report blocked time (s) | total validation time (s) | final score
| - | - | - | - | - | - |
| sync checkpointing + validation within training loop | 1451.03 | 6.09 | 14.71 | 164.87 | 0.58 |
| async checkpointing + validation with torch trainer | 1302.65 | 28.76 | 0.067 | 523.33 | 0.55 |
| async checkpointing + validation with map_batches | 1299.68 | 16.42 | 0.069 | 331.78 | 0.55 |

### CODE REVIEW

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    sync_validation_time = consolidated_metrics["sync_cp_inline_val_metrics"][
+        "total_validation_time"
+    ]
+    async_torchtrainer_validation_time = consolidated_metrics[
+        "async_cp_torch_trainer_val_metrics"
+    ]["total_validation_time"]
+    async_map_batches_validation_time = consolidated_metrics[
+        "async_cp_map_batches_val_metrics"
+    ]["total_validation_time"]
+    assert async_map_batches_validation_time < async_torchtrainer_validation_time
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: Why is map_batches necessarily faster than the torch trainer version? I think we should remove this assertion and just track as a release test metric

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    ]["e2e_time"]
+    assert (
+        async_torchtrainer_e2e_time
+        < sync_e2e_time * MAXIMUM_ALLOWED_E2E_TIME_MULTIPLIER
+        and async_map_batches_e2e_time
+        < sync_e2e_time * MAXIMUM_ALLOWED_E2E_TIME_MULTIPLIER
+    )
+
+    # Assert map_batches is faster than TorchTrainer. Note that inline is the fastest but is blocking
+    # Example values: 92s vs 387s vs 264s (gap between sync and async smaller if more data)
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:389]: why is the gap smaller with more data?

Comment (justinvyu) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:389]: Also, why is the difference so large? What's the extra 200 seconds of overhead coming from? Is this due to autoscaling time?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    model = create_model()
+    model = ray.train.torch.prepare_model(model)
+    criterion = CrossEntropyLoss()
+    optimizer = Adam(model.parameters(), lr=0.001)
+    train_data_shard = ray.train.get_dataset_shard("train")
+    train_dataloader = train_data_shard.iter_torch_batches(batch_size=256)
+    if validate_within_trainer:
+        test_dataloader = ray.train.get_dataset_shard("test").iter_torch_batches(
+            batch_size=128
+        )
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: I think we can simplify this test by removing the dataloading component -- just pass a dummy GPU tensor.

Comment (justinvyu) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: We can also mock out the model to just be a bunch of linear layers w/ a "model size" configuration

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            images, labels = batch["image"], batch["label"]
+            outputs = model(images)
+            mean_acc(outputs.argmax(1), labels)
+    ray.train.report(
+        metrics={"score": mean_acc.compute().item()},
+        checkpoint=ray.train.Checkpoint(
+            ray.train.get_context()
+            .get_storage()
+            .build_checkpoint_path_from_name("placeholder")
+        ),
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:119]: Let's just create the empty checkpoint similar to the the training train_fn

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+    )
+
+
+def validate_with_torch_trainer(checkpoint, config):
+    start_time = time.time()
+    trainer = ray.train.torch.TorchTrainer(
+        eval_only_train_func,
+        train_loop_config={"checkpoint": checkpoint},
+        scaling_config=ray.train.ScalingConfig(num_workers=2, use_gpu=True),
+        datasets={"test": config["dataset"]},
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:130]: I think this would be a good convention -- and pass these through the config

`name="{parent_run_name}-validation_epoch={x}_batch_idx={y}"`

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+        # Train model
+        model.train()
+        for batch in train_dataloader:
+            images, labels = batch["image"], batch["label"]
+            outputs = model(images)
+            loss = criterion(outputs, labels)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: can we test mid-epoch validation?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    row["image"] = transform(row["image"])
+    return row
+
+
+train_dataset = ray.data.read_parquet(f"{STORAGE_PATH}/cifar10-parquet/train").map(
+    transform_cifar
+)
+test_dataset = ray.data.read_parquet(f"{STORAGE_PATH}/cifar10-parquet/test").map(
+    transform_cifar
+)
```
Comment (justinvyu) [inline release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: move these out of the top level

**Review by justinvyu — COMMENTED** [review]
Thanks, looking great!

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 1 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: I thought it should be faster due to Ray Data optimizations. However I agree it's probably better to track as a metric than to assert.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 2 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:389]: The gap should be smaller with more data because the data iteration cost should amortize the validation startup cost. 

The validation startup cost includes:
* Scheduling the validation task, but this is not tracked (we only track time within the validation Ray task)
* Node autoscaling
* Scheduling the TorchTrainer workers or map_batches tasks
* Loading the checkpoint - I haven't tracked this, but checkpoint optimizations (e.g. GPU to GPU transfer) should make this better.

Added a TODO to track these better.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 3 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: What's the advantage of doing this over a more "realistic" benchmark like what this PR is currently doing?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 4 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:119]: I did it this way because:
* It does not write any information to disk or remote storage
* The long name is needed to avoid https://github.com/ray-project/ray/blob/4e3039eb402085eb79c06de6263b49e3255983a9/python/ray/train/v2/_internal/execution/storage.py#L543

Wdyt of just leaving it as is until we implement `train_func` return values and then replacing it with that?

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 5 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:130]: done.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 6 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: done.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 7 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: done.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    metrics["final_validation_waiting_time"] = (
+        end_time - train_func_metrics["train_func_return_time"]
+    )
+    metrics["total_report_blocked_time"] = sum(
+        train_func_metrics["report_blocked_times"]
+    )
+    metrics["total_validation_time"] = sum(
+        t[1]["validation_time"] for t in result.best_checkpoints[:num_epochs]
+    )
+    metrics["final_score"] = result.best_checkpoints[-2][1]["score"]
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:308]: fixed - thanks!

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+        for i, batch in enumerate(train_dataloader):
+            images, labels = batch["image"], batch["label"]
+            outputs = model(images)
+            loss = criterion(outputs, labels)
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            if i == midpoint_batch:
+                validate_and_report(model, epoch, i, blocked_times, config, loss)
+        validate_and_report(model, epoch, i, blocked_times, config, loss)
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:240]: shouldn't happen but worth protecting against - done.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+    def __init__(self, checkpoint):
+        self.model = create_model()
+        with checkpoint.as_directory() as checkpoint_dir:
+            model_state_dict = torch.load(os.path.join(checkpoint_dir, "model.pt"))
+            self.model.load_state_dict(model_state_dict)
+        self.model.cuda().eval()
+
+    def __call__(self, batch):
+        image = torch.as_tensor(batch["image"], dtype=torch.float32, device="cuda")
+        label = torch.as_tensor(batch["label"], dtype=torch.float32, device="cuda")
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: done.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+    if validate_within_trainer and epoch == num_epochs - 1:
+        metrics["score"] = mean_acc.compute().item()
+    if ray.train.get_context().get_world_rank() == 0:
+        if val_elapsed_time:
+            metrics["validation_time"] = val_elapsed_time
+        iteration_checkpoint_dir = tempfile.mkdtemp()
+        torch.save(
+            model.module.state_dict(),
+            os.path.join(iteration_checkpoint_dir, "model.pt"),
+        )
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:188]: report with async deletes the dir by default

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 3 above)
Comment (justinvyu) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: I mainly don't want this test to be tied to Ray Data, since then there's more moving pieces and Ray Data regressions could lead to this benchmark runtime increasing for unrelated reasons.

Keeping the model is fine actually. Would be good to have realistic GPU computation time.

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 4 above)
Comment (justinvyu) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:119]: I'm fine with uploading the empty directory similar to the train_fn

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 2 above)
Comment (justinvyu) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:389]: Oh gotcha makes sense.

**Review by justinvyu — APPROVED** [review]

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 4 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py:119]: leaving as is as discussed offline

File: release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py (hunk 3 above)
Comment (TimothySeah) [reply release/train_tests/async_checkpointing_validation_benchmark/test_async_checkpointing_validation_benchmark.py]: discussed offline - could be nice to test multiple datasets executing in same cluster
